src/ray/core_worker/core_worker.cc

```diff
@@ -131,6 +132,24 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
   });
   RAY_LOG(DEBUG) << "Constructing CoreWorker, worker_id: " << worker_id;
 
+  if (RayConfig::instance().kill_child_processes_on_worker_exit_with_raylet_subreaper()) {
+#ifdef __linux__
+    // Not setting sigchld = ignore: user may want to do waitpid on their own.
+    // If user's bad code causes a zombie process, it will hang their in zombie status
+    // until this worker exits and raylet reaps it.
+    if (SetThisProcessAsSubreaper()) {
+      RAY_LOG(INFO) << "Set this core_worker process as subreaper: " << getpid();
+    } else {
```

Comment:
When `kill_child_processes_on_worker_exit_with_raylet_subreaper` is set, why do we still need to set the core worker ? The raylet should just work as the reaper I think?

Reply:
See https://github.com/ray-project/ray/pull/42992/files#diff-f6c49babcc6278c29b15f311f28011254c7e7d20a0ff06b03a1ac17a6fe352bbR103

tldr: if a child of core worker spawns a grandchild and dies, the grandchild is reparented to the nearest subreaper. But we don't want it to be killed (since the core worker is still alive), so we set the core worker as a subreaper as well to avoid it be noticed and killed by raylet.

Reply:
Hmm, let me know if I understand it correctly:

- if the core worker is dead => the grand child reparented to the raylet
- if the raylet is dead => the core worker is still alive => later it'll kill itself => before it exits, it'll kill all the children?

Reply:
it handles this case:

raylet -> core worker -> A -> B

then

A exited, we don't want to kill B.

Now, B is reparented to core worker who does NOT kill B, so it's good. If we don't do that, B'd be reparented to raylet and be killed, which is what we don't want

Reply:
Thanks!